SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp

```diff
@@ -0,0 +1,254 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file provides functions for tests on simd decrement and increment
+/// operators.
+///
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "common.hpp"
+#include "../mutator.hpp"
+
+namespace esimd = sycl::ext::intel::experimental::esimd;
+namespace esimd_functional = esimd_test::api::functional;
+
+namespace esimd_test::api::functional::operators {
+
+// Descriptor class for the case of calling constructor in initializer context.
+struct pre_decrement {
+  static std::string get_operator_type() { return " pre decrement"; }
+
+  template <typename DataT, int NumElems>
+  static void call_simd_ctor(const DataT *const ref_data,
+                             DataT *const source_simd_out,
+                             DataT *const result_simd_out) {
+    auto source_simd = esimd::simd<DataT, NumElems>();
+    source_simd.copy_from(ref_data);
+    esimd::simd<DataT, NumElems> result_simd = --source_simd;
+    source_simd.copy_to(source_simd_out);
+    result_simd.copy_to(result_simd_out);
+  }
+
+  template <typename DataT> static DataT apply_operator(DataT val) {
+    return --val;
+  }
+
+  static constexpr bool is_increment() { return false; }
+};
+
+// Descriptor class for the case of calling constructor in initializer context.
+struct post_decrement {
+  static std::string get_operator_type() { return " post decrement"; }
+
+  template <typename DataT, int NumElems>
+  static void call_simd_ctor(const DataT *const ref_data,
+                             DataT *const source_simd_out,
+                             DataT *const result_simd_out) {
+    auto source_simd = esimd::simd<DataT, NumElems>();
+    source_simd.copy_from(ref_data);
+    esimd::simd<DataT, NumElems> result_simd = source_simd--;
+    source_simd.copy_to(source_simd_out);
+    result_simd.copy_to(result_simd_out);
+  }
+
+  template <typename DataT> static DataT apply_operator(DataT val) {
+    return val;
+  }
+
+  static constexpr bool is_increment() { return false; }
+};
+
+// Descriptor class for the case of calling constructor in initializer context.
+struct pre_increment {
+  static std::string get_operator_type() { return " pre increment"; }
+
+  template <typename DataT, int NumElems>
+  static void call_simd_ctor(const DataT *const ref_data,
+                             DataT *const source_simd_out,
+                             DataT *const result_simd_out) {
+    auto source_simd = esimd::simd<DataT, NumElems>();
+    source_simd.copy_from(ref_data);
+    esimd::simd<DataT, NumElems> result_simd = ++source_simd;
+    source_simd.copy_to(source_simd_out);
+    result_simd.copy_to(result_simd_out);
+  }
+
+  template <typename DataT> static DataT apply_operator(DataT val) {
+    return ++val;
+  }
```

Comment:
```suggestion
  template <typename DataT> static DataT apply_operator(DataT& val) {
    return ++val;
  }
```
By doing so you would be able:

- to remove `is_increment()` entirely
- to simplify "expected values" logic within `run_verification`
- to have exactly the post-increment/decrement operator within `apply_operator` for post-increment/decrement test case

For example:
```
DataT expected_source_value = shared_ref_data[i];
const DataT expected_return_value = TestCaseT::apply_operator(expected_source_value );
```

Reply:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/827/commits/07503714b6bf2d29f8318597e51142bec03a1fb8#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eL205-R210)